Issue #2426: Implement MAD Aggregate

hawkfish opened the pull request:
The Moving Absolute Deviation (or MAD) function is to medians what standard deviation is to average.  This CL implements the simple and moving versions of the aggregate. It also moves the quantile benchmark tests from the slow test suite to the benchmark suite. Testing shows that the moving version achieves a 3.5x performance gain (29.5s vs 8.5s) on 10M rows with 100 element fixed frames.

`MAD` is defined on temporal types as well as numeric types, which required adding a few methods to `Interval` to handle exact intervals more elegantly.

Commit:
Merge branch 'master' of github.com:duckdb/duckdb into hawkfish-reverse-precentile

Commit:
Issue #2425: Implement MAD

Refactor the quantile interpolation logic
so we can compile a MAD aggregate implementation.
Not tested.

Commit:
Issue #2425: Implement MAD

Fix amalgamation errors by factoring out AbsOperator.

Commit:
Issue #2425: Implement MAD

Remove tabs...

Commit:
Issue #2425: Implement MAD

Remove semicolon

Commit:
Issue #2425: Implement MAD

Register function and add tests for correctness and coverage.

Commit:
Issue #2425: Implement MAD

Clean up temporal rounding issues.

Commit:
Merge branch 'master' into hawkfish-mad

Commit:
Issue #2425: Implement MAD

Use exact, normalised INTERVALs for intelligibility.

Commit:
Merge branch 'master' into hawkfish-mad

Commit:
Merge branch 'master' into hawkfish-mad

Commit:
Issue #2425: Implement MAD

Moving window baseline tests.

Commit:
Issue #2425: Implement Moving MAD

This required further cleanup of the quantile computation templates.

Commit:
Issue #2425: Implement Moving MAD

Create benchmarks for 100 element frames
Also, start pulling out slow mode and quantile tests
into the benchmark suite.

Commit:
Merge branch 'hawkfish-mad' of https://github.com/hawkfish/duckdb into hawkfish-mad

Commit:
Merge branch 'master' into hawkfish-mad

Commit:
Issue #2425: Implement Moving MAD

Transfer moving IQR performance tests to the benchmark suite.

Commit:
Issue #2425: Implement MAD

Add ReplaceIndex coverage.

Commit:
Merge branch 'master' into hawkfish-mad

Mytherin reviewed (commented):
Thanks for the PR! Cool stuff. Some comments. Otherwise I would like to see the code coverage, but that is dependent on #2491 being merged.

Commit:
Issue #2425: Implement Moving MAD

Be more careful about extreme differences.

Commit:
Merge branch 'master' into hawkfish-mad

Commit:
Issue #2425: Implement Moving MAD

Resture use of Replacement for scalar QUANTILE and MAD.

Benchmark improvements:
* window_median_fixed_100: 3.0s => 3.45s
* window_median_variable_100: 9.0s => 8.7s
* window_mad_fixed_100: 8.25s => 7.42s
* window_mad_variable_100: Unchanged

Mytherin reviewed (approved):
Everything looks good now, thanks!

Mytherin reviewed (commented):
Thanks for the updates! Looks good to me, could you just have a look at the uncovered lines?

Commit:
Merge branch 'master' into hawkfish-mad

hawkfish commented on src/function/aggregate/holistic/quantile.cpp:
```diff
@@ -105,34 +132,32 @@ static idx_t ReplaceIndex(idx_t *index, const FrameBounds &frame, const FrameBou
+	const auto ik0 = index[k0];
+	const auto ik1 = index[k1];
+	if (!validity(ij) && !validity(ik0)) {
+		return true;
```
It was a good thing I took another look here because the `NULL` handling was still not correct.

Commit:
Issue #2425: Implement Moving MAD

Add coverage tests and fix NULL handling.
Also restore Operation/Replace regression
for quantile lists.

Mytherin commented:
Thanks!
